parsers/VN.py

```diff
@@ -1,216 +1,167 @@
 # Access via day, can also parse historical data
 
-tz = pytz.timezone("Asia/Ho_Chi_Minh")
-base_url_from_type = {
-    "consumption": "https://www.nldc.evn.vn/PhuTaiHandle.ashx?d=",  # + dd/mm/yyyy
-    "price": "https://www.nldc.evn.vn/GiaBienHandle.ashx?d=",  # + dd/mm/yyyy
+tz = zoneinfo.ZoneInfo("Asia/Ho_Chi_Minh")
+LIVE_DATA = {
+    "consumption": "https://www.nldc.evn.vn/api/services/app/Pages/GetChartPhuTaiVM",
+    "price": "https://www.nldc.evn.vn/api/services/app/Pages/GetChartGiaBienVM",
 }
 
+HISTORICAL_DATA = "https://www.nldc.evn.vn/api/services/app/Dashboard/GetBieuDoTuongQuanPT"  # + dd/mm/yyyy
 
-def fetch_latest_data(
-    data_type: str,
-    session: Session,
-):
-    """
-    Fetch latest available
-    """
-    fetched_day = arrow.now(tz=tz)
-
-    base_url = base_url_from_type[data_type]
-
-    today_dt_str = fetched_day.format("DD/MM/YYYY")
-
-    # Try fetch data for today.
-    today_response: Response = session.get(base_url + today_dt_str)
-
-    # Check if today's data is there.
-    if today_response.ok and today_response.json() is not None:
-        return fetched_day.floor("day"), today_response.json()
-
-    # If it is too early in the day, it might not contain data yet. Get from yesterday.
-    fetched_day = fetched_day.shift(days=-1)
-    yesterday_dt_str = fetched_day.format("DD/MM/YYYY")
-
-    yesterday_response: Response = session.get(base_url + yesterday_dt_str)
-    # Check if yesterday's data is there.
-    if yesterday_response.ok and yesterday_response.json() is not None:
-        return fetched_day.floor("day"), yesterday_response.json()
-
-    # Also no data from yesterday, cant fetch recent data.
-    raise ParserException(
-        parser="VN.py",
-        message=(
-            f"Requested recent data - request endpoint can't provide data for 2 days."
-            f"Return codes: {today_response.status_code}, {yesterday_response.status_code} at {fetched_day}"
-        ),
-    )
-
+ZONE_CODE = {ZoneKey("VN"): "HT"}
 
-def fetch_target_data(data_type: str, target_datetime, session: Session):
-    """
-    Fetch the day containing the given target_datetime.
-    """
 
-    # Convert timezone, and get the day
-    target_datetime_vn = arrow.get(target_datetime).to(tz)
+def fetch_consumption(
+    zone_key: ZoneKey,
+    session: Session = Session(),
+    target_datetime: datetime | None = None,
+    logger: Logger = getLogger(__name__),
+):
+    if target_datetime is None:
+        live_consumption = fetch_live_consumption(session, logger)
+        # Live data is empty at 00:00, resort to historical from previous day (which includes 00:00)
+        if len(live_consumption) > 0:
+            return live_consumption
+        else:
+            return fetch_historical_consumption(
+                datetime.now().astimezone(tz) - timedelta(days=1), session, logger
+            )
+    else:
+        return fetch_historical_consumption(target_datetime, session, logger)
```

Comment:
Why is there a astimezone on the first statement but not the seccond?

if you just want to set the timezone you can do that directly with datetime.now(tz=tz)